jax/_src/pallas/pallas_call.py

```diff
@@ -1063,6 +1065,9 @@ def pallas_call(
   if grid_spec is None:
     grid_spec = GridSpec(grid, in_specs, out_specs)
   grid_spec, dynamic_grid_bounds = grid_spec.unzip_dynamic_grid_bounds()
+  # TODO(necula): this canonicalization may be convenient for some usage
+  # but it is lossy, because it prevents expressing functions that return
+  # lists.
```

Comment:
I think it's common for many JAX functions to do it?

Reply:
Is it? I think only for the inputs, because the inputs are always a tuple, but I don't think we do this for outputs, or at least not for the cases where we determine whether the output is a tuple or a list. (E.g., in `pjit` the output container is determined by the user when they write the function, and the `out_shardings` simply annotation the outputs; but `out_shape` actually determines the container.)

See https://github.com/search?q=repo%3Agoogle%2Fjax%20https%3A%2F%2Fgithub.com%2Fgoogle%2Fjax%2Fissues%2F2367&type=code